Source/Core/VideoCommon/HiresTextures.cpp

```diff
@@ -238,10 +265,73 @@ std::string HiresTexture::GenBaseName(const u8* texture, size_t texture_size, co
 	return name;
 }
 
-HiresTexture* HiresTexture::Search(const u8* texture, size_t texture_size, const u8* tlut, size_t tlut_size, u32 width, u32 height, int format, bool has_mipmaps)
+std::shared_ptr<HiresTexture> HiresTexture::Search(const u8* texture, size_t texture_size, const u8* tlut, size_t tlut_size, u32 width, u32 height, int format, bool has_mipmaps)
 {
 	std::string base_filename = GenBaseName(texture, texture_size, tlut, tlut_size, width, height, format, has_mipmaps);
 
+	std::lock_guard<std::mutex> lk(s_textureCacheMutex);
+
+	auto iter = s_textureCache.find(base_filename);
+	if (iter != s_textureCache.end())
+	{
+		return iter->second;
+	}
+
+	std::shared_ptr<HiresTexture> ptr(Load(base_filename, width, height));
+
+	if (ptr && g_ActiveConfig.bCacheHiresTextures)
+	{
+		s_textureCache[base_filename] = ptr;
+	}
+
+	return ptr;
+}
+
+void HiresTexture::Prefetch()
+{
+	u32 starttime = Common::Timer::GetTimeMs();
+	for (const auto& entry : s_textureMap)
+	{
+		const std::string& base_filename = entry.first;
+
+		if (base_filename.find("_mip") == std::string::npos)
+		{
+			std::unique_lock<std::mutex> lk(s_textureCacheMutex);
+
+			auto iter = s_textureCache.find(base_filename);
+			if (iter == s_textureCache.end())
+			{
+				// unlock while loading a texture. This may result in a race condition where we'll load a texture twice,
+				// but it reduces the stuttering a lot. Notice: The loading library _must_ be thread safe now.
+				// But bad luck, SOIL isn't, so TODO: remove SOIL usage here and use libpng directly
+				//lk.unlock();
```

Comment:
Yes, this new comment is already outdated as explained in the last line. I did choose this comment to describe what should happen here through it is disabled right now. I hope this will be fixed by PR #2163